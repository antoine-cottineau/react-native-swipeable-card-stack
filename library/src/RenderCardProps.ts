import { type SharedValue } from 'react-native-reanimated'

export type RenderCardProps<T> = T & RenderCardAddedProps

export type RenderCardAddedProps = {
  index: number
  animationPosition: SharedValue<number>
  currentIndex: number
}
