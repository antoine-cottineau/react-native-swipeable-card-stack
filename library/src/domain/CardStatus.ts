export type CardStatus = 'swiped' | 'current' | 'unswiped'

type GetStatusParams = {
  index: number
  currentIndex: number
}

export const getStatus = ({
  index,
  currentIndex,
}: GetStatusParams): CardStatus => {
  if (index < currentIndex) {
    return 'swiped'
  }
  if (index === currentIndex) {
    return 'current'
  }
  return 'unswiped'
}
