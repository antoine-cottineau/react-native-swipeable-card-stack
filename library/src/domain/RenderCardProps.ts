import { type SharedValue } from 'react-native-reanimated'

export type RenderCardProps<T> = T & RenderCardAddedProps

export type RenderCardAddedProps = {
  /**
   * The index of the card in the stack.
   */
  index: number

  /**
   * The index of the card that is currently on the top of the stack.
   */
  currentIndex: number

  /**
   * A reanimated [SharedValue](https://docs.swmansion.com/react-native-reanimated/docs/core/useSharedValue/) that is synced with the swipe x position of the current card and can be used to perform custom animations.
   *
   * The value runs between -1 (the card is totally swiped to the left) and 1 (the card is totally swiped to the right).
   *
   * A value of 0 means the card is at its resting x position.
   */
  xAnimationPosition: SharedValue<number>

  /**
   * A reanimated [SharedValue](https://docs.swmansion.com/react-native-reanimated/docs/core/useSharedValue/) that is synced with the swipe y position of the current card and can be used to perform custom animations.
   *
   * The value runs between -1 (the card is totally swiped to the top) and 1 (the card is totally swiped to the bottom).
   *
   * A value of 0 means the card is at its resting y position.
   */
  yAnimationPosition: SharedValue<number>
}
