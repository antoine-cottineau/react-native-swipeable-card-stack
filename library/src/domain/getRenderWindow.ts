type Params = {
  currentIndex: number
  dataLength: number
  numberOfRenderedCards: number
}

/**
 * Get the window of indexes that should have their card rendered.
 *
 * Here are the rules that this function follows:
 * - the card at `currentIndex` must be rendered
 * - `numberOfRenderedCards` unswiped cards should be rendered
 * - if `currentIndex >= 1`, one swiped card should be rendered
 * - cards with negative indexes or with indexes superior to `dataLength` shouldn't be rendered
 *
 * @returns a tuple corresponding to the range of card indexes that should be rendered (inclusive).
 */
export const getRenderWindow = ({
  currentIndex,
  dataLength,
  numberOfRenderedCards,
}: Params): [number, number] => [
  Math.max(currentIndex - 1, 0),
  Math.min(currentIndex + numberOfRenderedCards, dataLength - 1),
]
