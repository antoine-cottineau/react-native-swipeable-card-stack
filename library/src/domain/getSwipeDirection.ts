import { type SwipeDirection } from './SwipeDirection'

type Params = {
  xTranslation: number
  yTranslation: number
  xEndedSwipePosition: number
  yEndedSwipePosition: number
}

export const getSwipeDirection = ({
  xTranslation,
  yTranslation,
  xEndedSwipePosition,
  yEndedSwipePosition,
}: Params): SwipeDirection => {
  'worklet'
  const horizontalRatio = Math.abs(xTranslation / xEndedSwipePosition)
  const verticalRatio = Math.abs(yTranslation / yEndedSwipePosition)

  if (horizontalRatio >= verticalRatio) {
    return xTranslation >= 0 ? 'right' : 'left'
  }
  return yTranslation >= 0 ? 'bottom' : 'top'
}
