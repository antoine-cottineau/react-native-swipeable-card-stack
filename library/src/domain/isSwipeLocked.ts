import { type SwipeAxis } from './SwipeAxis'
import { type SwipeDirection } from './SwipeDirection'

type Params = {
  translation: number
  axis: SwipeAxis
  lockedDirections: SwipeDirection[]
}

export const isSwipeLocked = ({
  translation,
  axis,
  lockedDirections,
}: Params): boolean => {
  const isXSwipeLocked =
    (lockedDirections.includes('left') && translation < 0) ||
    (lockedDirections.includes('right') && translation > 0)
  const isYSwipeLocked =
    (lockedDirections.includes('top') && translation < 0) ||
    (lockedDirections.includes('bottom') && translation > 0)

  return (axis === 'x' && isXSwipeLocked) || (axis === 'y' && isYSwipeLocked)
}
