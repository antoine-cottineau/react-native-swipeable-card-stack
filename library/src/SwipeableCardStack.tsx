import styled from '@emotion/native'
import {
  forwardRef,
  useImperativeHandle,
  useState,
  type ForwardedRef,
  type ReactNode,
} from 'react'
import { type StyleProp, type ViewStyle } from 'react-native'
import { useSharedValue } from 'react-native-reanimated'
import { type RenderCardProps } from './RenderCardProps'
import { type SwipeUpdate } from './SwipeUpdate'
import { type SwipeableCardStackOptions } from './SwipeableCardStackOptions'
import { SwipeableCardWrapper } from './SwipeableCardWrapper'
import { toReversed } from './toReversed'
import { useDefaultOptions } from './useDefaultOptions'
import { useRefMap } from './useRefMap'

export type SwipeableCardStackProps<T> = {
  data: T[]
  renderCard: (params: RenderCardProps<T>) => ReactNode
  cardWrapperStyle?: StyleProp<ViewStyle>
  onActiveCardUpdate?: (swipeUpdate: SwipeUpdate<T>) => void
} & Partial<SwipeableCardStackOptions>

export type SwipeableCardRef = {
  swipeLeft: () => void
  swipeRight: () => void
}

export const SwipeableCardStack = forwardRef(function SwipeableCardStack<T>(
  {
    data,
    renderCard,
    cardWrapperStyle,
    onActiveCardUpdate,
    ...userOptions
  }: SwipeableCardStackProps<T>,
  ref: ForwardedRef<SwipeableCardRef>,
) {
  const [currentIndex, setCurrentIndex] = useState(0)
  const refMap = useRefMap<number, SwipeableCardRef>()
  const animationPosition = useSharedValue(0)
  const defaultOptions = useDefaultOptions()

  const options: SwipeableCardStackOptions = {
    ...userOptions,
    ...defaultOptions,
  }

  useImperativeHandle(
    ref,
    () => ({
      swipeLeft: () => {
        refMap.get(currentIndex)?.swipeLeft()
      },
      swipeRight: () => {
        refMap.get(currentIndex)?.swipeRight()
      },
    }),
    [refMap, currentIndex],
  )

  return (
    <Container>
      {toReversed(data).map((cardData, reverseIndex) => {
        const index = data.length - reverseIndex - 1

        const hasCardBeenSwiped = index < currentIndex
        const shouldNotRenderCardYet =
          index > currentIndex + options.numberOfRenderedCards - 1
        if (hasCardBeenSwiped || shouldNotRenderCardYet) {
          return null
        }

        return (
          <SwipeableCardWrapper
            ref={(ref) => {
              refMap.initRef(index, ref)
            }}
            renderCard={(renderCardAddedProps) =>
              renderCard({ ...cardData, ...renderCardAddedProps })
            }
            index={index}
            animationPosition={animationPosition}
            currentIndex={currentIndex}
            cardWrapperStyle={cardWrapperStyle}
            onCardSwipeStatusUpdated={(swipeStatus) => {
              if (swipeStatus.phase === 'ended') {
                setCurrentIndex((index) => index + 1)
                animationPosition.value = 0
              }
              if (index === currentIndex) {
                const currentDataItem = data[currentIndex]
                if (currentDataItem === undefined) {
                  throw new Error(
                    `Attempted to access data item at index ${currentIndex} while data has size ${data.length}.`,
                  )
                }
                onActiveCardUpdate?.({
                  ...swipeStatus,
                  currentIndex,
                  currentDataItem,
                })
              }
            }}
            options={options}
            key={index}
          />
        )
      })}
    </Container>
  )
})

const Container = styled.View({
  flex: 1,
})
