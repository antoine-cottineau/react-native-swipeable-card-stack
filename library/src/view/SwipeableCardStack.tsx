import {
  forwardRef,
  memo,
  useImperativeHandle,
  useState,
  type ForwardedRef,
} from 'react'
import { StyleSheet, View } from 'react-native'
import { type SwipeableCardRef, type SwipeableCardStackProps } from '..'
import { getStatus } from '../domain/CardStatus'
import { getRenderWindow } from '../domain/getRenderWindow'
import { toReversed } from '../utils/toReversed'
import { useRefMap } from '../utils/useRefMap'
import { type SwipeableCardStackOptions } from './SwipeableCardStackOptions'
import { SwipeableCardWrapper } from './SwipeableCardWrapper'
import { useDefaultOptions } from './useDefaultOptions'
import { useSwipedCardsStore } from './useSwipedCardsStore'

const SwipeableCardStackMemo = forwardRef(function SwipeableCardStack<T>(
  {
    data,
    renderCard,
    cardWrapperStyle,
    onActiveCardUpdate,
    ...userOptions
  }: SwipeableCardStackProps<T>,
  ref: ForwardedRef<SwipeableCardRef>,
) {
  const [currentIndex, setCurrentIndex] = useState(0)
  const refMap = useRefMap<number, SwipeableCardRef>()
  const defaultOptions = useDefaultOptions()
  const { getSwipedCardSwipeDirection, pushSwipedCard, popSwipedCard } =
    useSwipedCardsStore()

  const options: SwipeableCardStackOptions = {
    ...defaultOptions,
    ...userOptions,
  }
  const [renderWindowMinIndex, renderWindowMaxIndex] = getRenderWindow({
    currentIndex,
    dataLength: data.length,
    numberOfRenderedCards: options.numberOfRenderedCards,
  })

  useImperativeHandle(
    ref,
    () => ({
      swipe: (direction) => {
        refMap.get(currentIndex)?.swipe(direction)
      },
      unswipe: () => {
        if (currentIndex === 0) {
          return
        }
        setCurrentIndex((index) => index - 1)
        refMap.get(currentIndex - 1)?.unswipe()
        popSwipedCard()
      },
    }),
    [refMap, currentIndex, popSwipedCard],
  )

  return (
    <View style={styles.container}>
      {toReversed(data).map((cardData, reverseIndex) => {
        const index = data.length - reverseIndex - 1

        if (index < renderWindowMinIndex || index > renderWindowMaxIndex) {
          return null
        }

        return (
          <SwipeableCardWrapper
            ref={(ref) => {
              refMap.initRef(index, ref)
            }}
            renderCard={(renderCardAddedProps) =>
              renderCard({ data: cardData, ...renderCardAddedProps })
            }
            index={index}
            status={getStatus({ index, currentIndex })}
            cardWrapperStyle={cardWrapperStyle}
            onCardSwipeStatusUpdated={(swipeStatus) => {
              if (swipeStatus.phase === 'ended') {
                setCurrentIndex((index) => index + 1)
                pushSwipedCard(swipeStatus.direction)
              }
              if (index === currentIndex) {
                const currentDataItem = data[currentIndex]
                if (currentDataItem === undefined) {
                  throw new Error(
                    `Attempted to access data item at index ${currentIndex} while data has size ${data.length}.`,
                  )
                }
                onActiveCardUpdate?.({
                  ...swipeStatus,
                  currentIndex,
                  currentDataItem,
                })
              }
            }}
            options={options}
            initialSwipeDirection={getSwipedCardSwipeDirection(index)}
            key={index}
          />
        )
      })}
    </View>
  )
})

export const SwipeableCardStack = memo(
  SwipeableCardStackMemo,
) as typeof SwipeableCardStackMemo

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
})
