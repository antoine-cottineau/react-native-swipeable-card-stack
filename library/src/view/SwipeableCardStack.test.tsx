import { render, screen } from '@testing-library/react-native'
import { type ReactNode } from 'react'
import { View, Text } from 'react-native'
import { SafeAreaProvider } from 'react-native-safe-area-context'
import { type RenderCardProps } from '../domain/RenderCardProps'
import { SwipeableCardStack } from './SwipeableCardStack'

type FruitCardProps = RenderCardProps<{
  fruit: string
}>

const FruitCard = ({ fruit }: FruitCardProps) => (
  <View>
    <Text>{fruit}</Text>
  </View>
)

it('correctly renders', () => {
  render(
    <SwipeableCardStack
      data={[
        { fruit: 'banana' },
        { fruit: 'apple' },
        { fruit: 'lemon' },
        { fruit: 'strawberry' },
      ]}
      renderCard={FruitCard}
    />,
    { wrapper: TestWrapper },
  )

  expect(screen.getByText('banana')).toBeOnTheScreen()
})

const TestWrapper = ({ children }: { children: ReactNode }) => (
  <SafeAreaProvider>{children}</SafeAreaProvider>
)
