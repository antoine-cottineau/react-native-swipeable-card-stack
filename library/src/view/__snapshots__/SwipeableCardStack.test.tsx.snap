// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly renders 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 0,
            },
            {
              "translateY": 0,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "height": "100%",
        "position": "absolute",
        "transform": [
          {
            "translateX": 0,
          },
          {
            "translateY": 0,
          },
        ],
        "width": "100%",
      }
    }
    testID="swipeable-card-wrapper-3"
  >
    <View
      collapsable={false}
    >
      <Text>
        strawberry
      </Text>
    </View>
  </View>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 0,
            },
            {
              "translateY": 0,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "height": "100%",
        "position": "absolute",
        "transform": [
          {
            "translateX": 0,
          },
          {
            "translateY": 0,
          },
        ],
        "width": "100%",
      }
    }
    testID="swipeable-card-wrapper-2"
  >
    <View
      collapsable={false}
    >
      <Text>
        lemon
      </Text>
    </View>
  </View>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 0,
            },
            {
              "translateY": 0,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "height": "100%",
        "position": "absolute",
        "transform": [
          {
            "translateX": 0,
          },
          {
            "translateY": 0,
          },
        ],
        "width": "100%",
      }
    }
    testID="swipeable-card-wrapper-1"
  >
    <View
      collapsable={false}
    >
      <Text>
        apple
      </Text>
    </View>
  </View>
  <View
    animatedStyle={
      {
        "value": {
          "transform": [
            {
              "translateX": 0,
            },
            {
              "translateY": 0,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      {
        "height": "100%",
        "position": "absolute",
        "transform": [
          {
            "translateX": 0,
          },
          {
            "translateY": 0,
          },
        ],
        "width": "100%",
      }
    }
    testID="swipeable-card-wrapper-0"
  >
    <View
      collapsable={false}
    >
      <Text>
        banana
      </Text>
    </View>
  </View>
</View>
`;
